spock-specs/src/test/groovy/org/spockframework/smoke/extension/TempDirExtensionSpec.groovy

```diff
@@ -0,0 +1,158 @@
+package org.spockframework.smoke.extension
+
+import org.spockframework.EmbeddedSpecification
+
+import java.nio.file.Files
+import java.nio.file.Path
+
+/**
+ * @author: dqyuan
+ * @date: 2020/07/12
+ */
+class TempDirExtensionSpec extends EmbeddedSpecification {
+
+  static List<Path> tmpPaths = []
+
+  def setup() {
+    tmpPaths.clear()
+  }
+
+  def assertTmpPathsNotExist() {
+    tmpPaths.each {
+      path -> assert !Files.exists(path)
+    }
+  }
+
+  def "@TempDir creates only one dir for one spec, if the annotated field is 'shared'"() {
+    when:
+    def result = runner.runWithImports("""import spock.lang.Shared
+import spock.lang.Stepwise
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+
+@Stepwise
+class Foo extends Specification {
+
+  @Shared
+  @TempDir
+  Path dir1
+
+  @Shared
+  @TempDir
+  File dir2
+
+  static String testContent = UUID.randomUUID().toString()
+
+  static String subFile = "dirtest.txt"
+
+  def test1() {
+    expect:
+    Path testTxt = dir1.resolve(subFile)
+    Files.write(testTxt, testContent.getBytes())
+    File testTxt2 = new File(dir2, subFile)
+    Files.write(Paths.get(testTxt2.toURI()), testContent.getBytes())
+    org.spockframework.smoke.extension.TempDirExtensionSpec.tmpPaths.add(dir1)
+    org.spockframework.smoke.extension.TempDirExtensionSpec.tmpPaths.add(Paths.get(dir2.toURI()))
+  }
+
+  def test2() {
+    when:
+    def txt1 = Files.readAllLines(dir1.resolve(subFile))[0]
+    def txt2 = Files.readAllLines(Paths.get(dir2.toURI()).resolve(subFile))[0]
+
+    then:
+    txt1 == testContent
+    txt2 == testContent
+  }
+
+}
+""")
+
+    then:
+    result.testsStartedCount == 2
+    result.testsSucceededCount == 2
+    assertTmpPathsNotExist()
+  }
+
+  def "@TempDir creates one temp dir for one feature method, if the annotated field is not 'shared'"() {
```

Comment:
You should rewrite this test as data-driven test without the `runner`.
Then you do not need a `static` field, do not need a nested specification and also test that each **iteration** gets its own temp directory. You can e. g. use a shared field instead and verify that the dir of the previous iteration was removed and that you have a new one in the second iteration.